### DISCUSSION TITLE
How to Handle File Operations

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#1450] [Q&A] [answered]
My editor is hosted in the browser with the language server running in a web worker so I am wondering how should we go about handling file operations in the language server where you are outside of the Node context. 

We could send LSP messages from the client in the browser to the server in the worker, but I am wondering if this is already handled for us in Langium v3.0.0 using the workspace.  My guess is that I should  implement a custom `FileSystemProvider` that can provide the linkage between the browser-managed file system and the `DefaultWorkspaceManager`.  My goal is to synchronize the create/open/rename/delete/close operations for files in our editing environment.

Am I thinking about this correctly?

**Comment by msujew** [comment]
Hey @ballcoach12,

open/change operations should be handled by the language server by default using the appropriate editor LSP messages. For creation/deletion and rename (which is effectively just a deletion + creation), you will need the [`workspace/didChangeWatchedFiles`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles) notification. I assume that you can hook into the language client you're using (monaco-languageclient?) and send the notification yourself if you register a change in the workspace.

Note that the language server will then attempt to load the created document, so you need a browser based implementation of `FileSystemProvider` that is capable of loading the file somehow across the webworker border.

**Reply by ballcoach12** [reply]
Thanks, @msujew.  Sounds like I am heading in the right direction.   

One more question: we are currently setting the editor's code using the `setMainCode()` method on a `CodeEditorConfig` instance from monaco-editor-wrapper.  Doing it this way results in a self-assigned URI for the parsed document in the language server; so that is essentially how we are handling the file/open operation.  It sounds like we need to replace this approach with the sending of the appropriate LSP message to the language server to open the file.  If we do this, though, I'm not sure how the code gets set in the editor.  Does this happen automatically as a result of sending that LSP message to open the file?

**Reply by msujew** [reply]
I believe the monaco-languageclient (under the monaco-editor-wrapper) will automatically send the `textDocument/open` notification to the server when the text model has been set. However, I'm not too deep in there and believe @kaisalmen is better suited to answer this question.

Kai, care to share your thoughts on this?

**Reply by cdietrich** [reply]
We use model refs client side 
https://github.com/cdietrich/my-monaco-editor-react-example/blob/cd/otherFilesExperiments_createModelReference_multi_update/src/MonacoEditorReactCompExtended.ts

with (https://github.com/cdietrich/langium-did-close-support)

**Reply by cdietrich** [reply]
But in you case I don’t I standing why setting the uri along with the main code is not sufficient

**Reply by ballcoach12** [reply]
@cdietrich, I think it would be sufficient for us.  I just don't see how we can set the URI of the document in the workspace; but maybe I am misunderstanding how this works.  My idea is to make the URI the same as the path for the file in the browser-based file system.

**Reply by cdietrich** [reply]
@ballcoach12 is this one not exposed in the plan wrapper api: https://github.com/cdietrich/my-monaco-editor-react-example/blob/bafbddbca74a3b93678bd8ae8fae1571859234de/src/App.tsx#L57

**Reply by ballcoach12** [reply]
@cdietrich, what version of `monaco-editor-wrapper` are you using in that example? I do not see it referenced in the package.json for that project.

**Reply by cdietrich** [reply]
You need to check the @typefox/monaco-editor-react module

**Reply by cdietrich** [reply]
53  "dependencies": {
54    "monaco-editor": "npm:@codingame/monaco-vscode-editor-api@~3.1.2",
55    "monaco-editor-wrapper": "~4.0.0",
56    "react": "~18.2.0",
57    "vscode": "npm:@codingame/monaco-vscode-api@~3.1.2"
58  },

**Reply by kaisalmen** [reply]
Hi @ballcoach12 this is probably the most helpful config from the `monaco-languageclient` project regarding the question: https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/config.ts

This configuration is used by the [wrapper example](https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/main.ts) and [react example](https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/reactPython.tsx). The config defines an `codeUri` and defines and sets the `workspaceFolder` accordingly in the client options. If the language server has a file at the specified uri you should get the content automatically.

**Reply by ballcoach12** [reply]
@kaisalmen, I assume that, as @msujew suggested, I'll need to register my own `FileSystemProvider` for this case?

**Reply by kaisalmen** [reply]
@ballcoach12 yes, that is also what is done in the linked wrapper example ([this one](https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/main.ts))

**Reply by ballcoach12** [reply]
Thanks, @kaisalmen.  I think I am a little confused about the methodology here.  I was under the impression that I'd inject my own `WorkspaceManager` and give it my `FileSystemProvider` implementation, with the `FileSystemProvider` implementation handling the processing of the workspace and the  `WorkspaceManager` sending the files to contents to the document  builder to get LangiumDocuments for the index.  But the example you've given seems to use a client-side approach that involves the Monaco editor.

Is there more formal documentation on how these services work so that I don't have to keep peppering you guys with questions? :)

**Reply by kaisalmen** [reply]
@ballcoach12 maybe I am confused here. 🙂 In the examples I mentioned, the client is responsible for providing the files and the language servers just reacts to that. In case a model/uri is opened in monaco-editor the language server gets notified/reacts to that. The client is the driver. I thought what you originally outlined should have been client driven as well.

An example that we currently don't yet have for the wrapper is a multi-file example demonstrating cross-references or imports. This could be beneficial for understanding, I guess.

**Reply by ballcoach12** [reply]
@kaisalmen I think I am trying to figure out the best approach for my design.  

Based on my understanding of what @msujew suggested, we'd just opt for the client send LSP messages to the language server, and let the language server manage retrieving documents and handling workspace operations.  But it sounds like workspace and file operations could be managed on either side. 

So if we went the route of letting the language server handle file operations (because we want to parse all the files in the workspace so that their AST elements are in scope) then my assumption is that setting `codeUri` on the configuration object would result in the appropriate LSP message being sent to the server.  Am I correct about that?

**Reply by kaisalmen** [reply]
@ballcoach12 yes, that's correct. When languageclient (LC) and languageserver (LS) (languageserver was from the start a misleading term. It is basically processor or compute kernel and not a server. My brain always links this term to something more complex) don't share the same file system, then you have to somehow ensure that both side see the same content. In contrast when you work locally in VSCode, that is usually not the case as both share the same file system. Then the LS is able to resolve other files in the same way the LC can.

In a distributed case you have to ensure this and select the side that drives it. In the monaco-editor-wrapper/monaco-languageclient case that driver should be the client. You can do something like this and the LS (python/pyright in this case) will pick it up (=you don't have to open all files in the editor, just announce the via vscode api). Underlying `monaco-vscode-api` supplies these capabilities, so you can make use of the vscode api:
```ts
const testUri = vscode.Uri.file('/workspace/test.py');
fileSystemProvider.registerFile(new RegisteredMemoryFile(testUri, `const hello = "Hello World";
console.log(hello);
`
));
const testPy = await vscode.workspace.openTextDocument(testUri);
console.log(testPy.getText());
```

**Reply by cdietrich** [reply]
btw this is the way i did it in the example mentioned before.
we open the to be edited file in the foreground and all other files (imported ones) in the background using monacos `createModelReference`

**Reply by kaisalmen** [reply]
@cdietrich using `vscode.workspace.openTextDocument` may be interesting for you as well

**Reply by kaisalmen** [reply]
@ballcoach12 and @cdietrich sorry for coming up with this proposal only now. I sometimes forget we have all those capabilities. 😅 I updated the [python client example with the bespoke code](https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/main.ts#L21-L46)

**Reply by kaisalmen** [reply, marked as answer]
> Thanks [@kaisalmen](https://github.com/kaisalmen). I think I am close to having this working with the latest m-e-w version. However, it looks like the editorWorker-es.js file can't be found. Is buildWorkerDefinition() still the correct way to load it? I noticed that your code example has a configureMonacoWorkers() function. Does that tell the worker factory where to find the worker js file?

I received this notification, but I don't see it here. 🤔 `buildWorkerDefinition` is outdated. I introduced `configureMonacoWorkers` with v4 of the wrapper. The new function is more flexible. You can use the bundle version like this: https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/python/client/main.ts#L14-L18 or fully override the instructions like here: https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/src/ts/wrapperTs.ts#L16-L23

**Reply by ballcoach12** [reply]
Sorry for the confusion, @kaisalmen.  After I made the post, I figured out the code changes that I needed to make and didn't want to waste anyone's time.  Thank you for the response, though.  It sheds even more light on how things work.
